### PR TITLE
feat: added access to the owner only function 'registerGatewayToken'

### PR DESCRIPTION
**Pull request by Foivos**
[AXE-2997]

[AXE-2997]: https://axelarnetwork.atlassian.net/browse/AXE-2997?atlOrigin=eyJpIjoiNWRkNTljNzYxNjVmNDY3MDlhMDU5Y2ZhYzA5YTRkZjUiLCJwIjoiZ2l0aHViLWNvbS1KU1cifQ

### CODE REVIEW

File: evm/interchainTokenFactory.js [hunk 1]
```diff
@@ -207,6 +209,43 @@ async function processCommand(config, chain, options) {
             break;
         }
 
+        case 'registerGatewayTokens': {
```
Comment (milapsheth) [inline evm/interchainTokenFactory.js:212]: also add the singular registerGatewayToken for completeness

File: evm/interchainTokenFactory.js [hunk 2]
```diff
@@ -207,6 +209,43 @@ async function processCommand(config, chain, options) {
...
+            let { assetApi, tokenInfoApi, batchSize } = options;
+
+            assetApi = assetApi || `${config.axelar.lcd}/axelar/nexus/v1beta1/assets/`;
+            tokenInfoApi = tokenInfoApi || `${config.axelar.lcd}/axelar/evm/v1beta1/token_info/`;
```
Comment (milapsheth) [inline evm/interchainTokenFactory.js:216]: get lcd base url from the config

File: evm/interchainTokenFactory.js [hunk 3]
```diff
@@ -207,6 +209,43 @@ async function processCommand(config, chain, options) {
...
+            batchSize = batchSize || 10;
+
+            validateParameters({ isString: { assetApi, tokenInfoApi }, isNumber: { batchSize } });
+
+            const { assets } = (await axios.get(`${assetApi}${chain.name}`)).data;
```
Comment (milapsheth) [inline evm/interchainTokenFactory.js:221]: we have a httpGet helper. see how we query axelar auth info in gateway deployment

File: evm/interchainTokenFactory.js [hunk 4]
```diff
@@ -207,6 +209,43 @@ async function processCommand(config, chain, options) {
...
+
+            for (let i = 0; i < assets.length; i += batchSize) {
+                const asset = assets[i];
+                const salt = keccak256(hexlify(toUtf8Bytes(asset)));
+                const tokenId = await interchainTokenService.interchainTokenId(AddressZero, salt);
+                const tokenManagerAddress = await interchainTokenService.tokenManagerAddress(tokenId);
+
+                if ((await provider.getCode(tokenManagerAddress)).length > 2) {
```
Comment (milapsheth) [inline evm/interchainTokenFactory.js:229]: the logic is fragile to changes in the batch size variable. split this into two separate loops. one that collects the list of all assets that aren't registered, and another that batch registers them based on the batch size. You can make it more functional by using map/filter.

File: evm/interchainTokenFactory.js [hunk 5]
```diff
@@ -207,6 +209,43 @@ async function processCommand(config, chain, options) {
...
+
+                for (let j = i; j < i + batchSize && j < assets.length; j++) {
+                    const asset = assets[j];
+                    const salt = keccak256(hexlify(toUtf8Bytes(asset)));
+                    const { symbol } = (await axios.get(`${tokenInfoApi}${chain.name}?asset=${asset}`)).data.details;
+                    const { data } = await interchainTokenFactory.populateTransaction.registerGatewayToken(salt, symbol);
+                    multicallData.push(data);
+                }
+
+                await (await interchainTokenFactory.multicall(multicallData)).wait();
```
Comment (milapsheth) [inline evm/interchainTokenFactory.js:243]: use `handleTx`. See other methods in this file

File: evm/interchainTokenFactory.js (hunk 1 above)
Comment (milapsheth) [inline evm/interchainTokenFactory.js:212]:
```suggestion
        case 'registerAllGatewayTokens': {
```
a bit more explicit

File: evm/interchainTokenFactory.js [hunk 6]
```diff
@@ -207,6 +209,46 @@ async function processCommand(config, chain, options) {
...
         }
 
+        case 'registerAllGatewayTokens': {
+            let { assetApi, tokenInfoApi, batchSize } = options;
+
+            assetApi = assetApi || `${config.axelar.lcd}/axelar/nexus/v1beta1/assets/`;
+            tokenInfoApi = tokenInfoApi || `${config.axelar.lcd}/axelar/evm/v1beta1/token_info/`;
+            batchSize = batchSize || 10;
+
+            validateParameters({ isString: { assetApi, tokenInfoApi }, isNumber: { batchSize } });
```
Comment (deanamiel) [inline evm/interchainTokenFactory.js:219]:
```suggestion
            validateParameters({ isNonEmptyString: { assetApi, tokenInfoApi }, isNumber: { batchSize } });
```

File: evm/interchainTokenFactory.js [hunk 7]
```diff
@@ -207,6 +209,46 @@ async function processCommand(config, chain, options) {
...
+
+            const { assets } = (await axios.get(`${assetApi}${chain.name}`)).data;
```
Comment (deanamiel) [inline evm/interchainTokenFactory.js:221]: Wrap in `try/catch` in case API call fails?

File: evm/interchainTokenFactory.js [hunk 8]
```diff
@@ -207,6 +209,46 @@ async function processCommand(config, chain, options) {
...
+                }
+            }
+
+            for (let i = 0; i < unregisteredAssets.length; i += batchSize) {
+                const multicallData = [];
+
+                for (let j = i; j < i + batchSize && j < unregisteredAssets.length; j++) {
+                    const asset = unregisteredAssets[j];
+                    const salt = keccak256(hexlify(toUtf8Bytes(asset)));
+                    const { symbol } = (await axios.get(`${tokenInfoApi}${chain.name}?asset=${asset}`)).data.details;
```
Comment (deanamiel) [inline evm/interchainTokenFactory.js:240]: Wrap in `try/catch` in case API call fails?

File: evm/interchainTokenFactory.js [hunk 9]
```diff
@@ -207,6 +209,46 @@ async function processCommand(config, chain, options) {
...
+                    const { data } = await interchainTokenFactory.populateTransaction.registerGatewayToken(salt, symbol);
+                    multicallData.push(data);
+                }
+
+                const tx = await interchainTokenFactory.multicall(multicallData, { gasOptions });
```
Comment (deanamiel) [inline evm/interchainTokenFactory.js:245]:
```suggestion
                const tx = await interchainTokenFactory.multicall(multicallData, gasOptions);
```
`getGasOptions` should return `gasOptions` as an object already

File: evm/interchainTokenFactory.js [hunk 10]
```diff
@@ -237,6 +279,7 @@ if (require.main === module) {
                 'deployRemoteInterchainToken',
                 'registerCanonicalInterchainToken',
                 'deployRemoteCanonicalInterchainToken',
+                'registerAllGatewayTokens',
```
Comment (deanamiel) [inline evm/interchainTokenFactory.js:282]: Make sure to add these variables as options below: `assetApi, tokenInfoApi, batchSize`

File: package.json [hunk 11]
```diff
@@ -25,7 +25,7 @@
     "@0xpolygonhermez/zkevm-commonjs": "github:0xpolygonhermez/zkevm-commonjs#v1.0.0",
     "@axelar-network/axelar-cgp-solidity": "6.2.1",
     "@axelar-network/axelar-gmp-sdk-solidity": "5.6.4",
-    "@axelar-network/interchain-token-service": "1.2.1",
+    "@axelar-network/interchain-token-service": "https://github.com/axelarnetwork/interchain-token-service.git",
```
Comment (deanamiel) [inline package.json:28]: We already discussed this, but once we have a stable release changing it here should fix the tests

**Pull request closed without merging**